# Homework 8
### Jiaxin Li 77664780

In this homework, please feel free to call the functions in sklean.

Basic Requirements:
1. load the hand-written digit data, and split train and test datasets (follow the last homework);
2. Fit the kNN and random forest model separately on training data, and report the performance on test data -- you don't have to tune the parameters for basic requirements, choosing the parameters as you like;
3. Compare the performace with logistic regression;
4. Write some comments in the code, OR use markdown paragraphs to indicate what you are doing in the specific step.

Optional Task:
Using cross-validations to select the parameters when fitting the model.

##### Step 1: Loading data and split train and test datasets

In [1]:
from sklearn.datasets import load_digits   #Import data
X,y = load_digits(return_X_y = True)      #Set X to be dataset that contains 1798 samples, and y to be lables for each row in X.
X.shape

(1797, 64)

Split datasets

In [2]:
import numpy as np
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=40)   #Randomly take out 20% of the dataset as testing data. 
print(X_train.shape)   
print(y_test.shape)

(1437, 64)
(360,)


###### Stpe 2: Fit the kNN and random forest model separately on training data, and report the performance on test data

Determine which class a sample come from by counting the nearist 10 samples, which label appear most times.

In [3]:
from sklearn.neighbors import KNeighborsClassifier
knn_clf = KNeighborsClassifier(n_neighbors = 10)    
knn_clf.fit(X_train, y_train)      
a=knn_clf.score(X_test,y_test)
print("the accuracy on determining lable on test dataset is", a*100, "%")

the accuracy on determining lable on test dataset is 97.5 %


 make 1000 decision trees by random picking up 90% of the dataset, and each tree has the maximum depth of 5 in order to avoid overfitting. njobs = -1 means you ask to use all the processors of your computer

In [4]:
from sklearn.ensemble import RandomForestClassifier
rf_clf = RandomForestClassifier(n_estimators=1000, max_samples = 0.5, max_depth=5, random_state=0, n_jobs = -1)
rf_clf.fit(X_train, y_train) # train the "forest" on training dataset. fit perimeters
b=rf_clf.score(X_test, y_test)  # test on testing datast
print("the accuracy on determining lable on test dataset is", b*100, "%")

the accuracy on determining lable on test dataset is 91.94444444444444 %


###### Stpe 3: Compare with logistic regression

In [31]:
from sklearn.linear_model import LogisticRegression
clf=LogisticRegression(max_iter=100000)
clf.fit(X_train,y_train)   #find the optimal coefficients
y_pred=clf.predict(X_test)
p=np.sum(y_pred==y_test)*1/len(y_pred)*100  #compare each entry in pred_y and y_test to see how many of them are the same. And then find the accuracy.
print("accuracy:", p,"%")

accuracy: 96.38888888888889 %


##### summary
from the results above we can see the KNN method perform the best with 97.5% accuracy, and Logistic regression the second with 96.389% accuracy. The lowest one is RandomForestClassifier, with accuracy 91.94%. 
### In general all three ways perform well on preditcting the lable on test dataset.

# Optional, using cross validation to determine k.

In [5]:
import pandas as pd
from sklearn.model_selection import cross_val_score

k_list = list(range(1,50,10))
# creating dataframe of cv scores and test scores -- of course you can also use Numpy array
cv_scores = pd.DataFrame()
test_scores = pd.Series(dtype = 'float64')

# perform 10-fold cross validation
for k in k_list:
    knn_clf.set_params(n_neighbors=k) # update the object
    scores = cross_val_score(knn_clf, X_train, y_train, cv=10, scoring='accuracy')
    cv_scores[str(k)] = scores
    test_scores[str(k)] = knn_clf.score(X_test,y_test)

from the table below, the top row 1, 11, 21, 31, 41 are the different k values, and left colums represent how each k perform on 10 different test dataset.

In [6]:
cv_scores

1        11        21        31        41
0  0.979167  0.958333  0.958333  0.951389  0.930556
1  0.979167  0.965278  0.958333  0.951389  0.944444
2  0.993056  0.979167  0.979167  0.972222  0.965278
3  0.993056  0.972222  0.958333  0.951389  0.923611
4  1.000000  0.986111  0.979167  0.979167  0.951389
5  0.993056  0.993056  0.986111  0.986111  0.986111
6  1.000000  0.993056  0.993056  0.986111  0.979167
7  0.979021  0.972028  0.965035  0.965035  0.965035
8  0.979021  0.979021  0.965035  0.951049  0.930070
9  0.986014  0.972028  0.951049  0.951049  0.930070

In [7]:
cv_scores.mean()

1     0.988156
11    0.977030
21    0.969362
31    0.964491
41    0.950573
dtype: float64

we can find that k=1 perform the best for the knn model.

In [8]:
knn_clf = KNeighborsClassifier(n_neighbors = 1)    
knn_clf.fit(X_train, y_train)      
a=knn_clf.score(X_test,y_test)
print("the accuracy on determining lable on test dataset is", a*100, "%")

the accuracy on determining lable on test dataset is 98.61111111111111 %
